tools/ci/pr_preview.py

```diff
@@ -0,0 +1,403 @@
+#!/usr/bin/env python
+
+# The service provided by this script is not critical, but it shares a GitHub
+# API request quota with critical services. For this reason, all requests to
+# the GitHub API are preceded by a "guard" which verifies that the subsequent
+# request will not deplete the shared quota.
+#
+# In effect, this script will fail rather than interfere with the operation of
+# critical services.
+
+import argparse
+import json
+import logging
+import os
+import subprocess
+import time
+
+import requests
+
+# The ratio of "requests remaining" to "total request quota" below which this
+# script should refuse to interact with the GitHub.com API
+API_RATE_LIMIT_THRESHOLD = 0.2
+# The GitHub Pull Request label which indicates that a Pull Request is expected
+# to be actively mirrored by the preview server
+LABEL = 'safe for preview'
+# The number of seconds to wait between attempts to verify that a submission
+# preview is available on the Pull Request preview server
+POLLING_PERIOD = 5
+# Pull Requests from authors with the following associations to the project
+# should automatically receive previews
+#
+# https://developer.github.com/v4/enum/commentauthorassociation/ (equivalent
+# documentation for the REST API was not available at the time of writing)
+TRUSTED_AUTHOR_ASSOCIATIONS = ('COLLABORATOR', 'MEMBER', 'OWNER')
+# These GitHub accounts are not associated with individuals, and the Pull
+# Requests they submit rarely require a preview.
+AUTOMATION_GITHUB_USERS = (
+    'autofoolip', 'chromium-wpt-export-bot', 'moz-wptsync-bot',
+    'servo-wpt-sync'
+)
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+def gh_request(method_name, url, body=None, media_type=None):
+    github_token = os.environ.get('GITHUB_TOKEN')
+
+    kwargs = {
+        'headers': {
+            'Authorization': 'token {}'.format(github_token),
+            'Accept': media_type or 'application/vnd.github.v3+json'
+        }
+    }
+    method = getattr(requests, method_name.lower())
+
+    if body is not None:
+        kwargs['json'] = body
+
+    logger.info('Issuing request: %s %s', method_name.upper(), url)
+
+    resp = method(url, **kwargs)
+
+    resp.raise_for_status()
+
+    logger.info('Response status code: %s', resp.status_code)
```

Comment:
Nit; we will never log non-successful status codes. (I assume, given `raise_for_status` above)

Reply:
The error message will describe the status, but a caller could handle or override the exception. I'll swap these two lines, just to be safe.